include/blast/math/algorithm/Trmm.hpp

```diff
@@ -67,159 +71,256 @@ namespace blast
-    requires Matrix<MTB, ET> && Matrix<MTA, ET> && Matrix<MTC, ET>
-        && (StorageOrder_v<MTB> == columnMajor) && (StorageOrder_v<MTC> == columnMajor)
-    inline void trmmRightLower(
-        ET alpha,
-        MTB const& B, MTA const& A,
-        MTC& C)
+    /// See https://netlib.org/lapack/explore-html-3.6.1/d1/d54/group__double__blas__level3_gaf07edfbb2d2077687522652c9e283e1e.html
+    ///
+    /// @tparam MPB matrix pointer type for matrix B
+    /// @tparam MPA matrix pointer type for matrix A
+    /// @tparam MPC matrix pointer type for matrix C
+    ///
+    /// @param M the number of rows of B
+    /// @param N the number of columns of B
+    /// @param alpha the scalar alpha
+    /// @param B pointer to top left element of matrix B
+    /// @param A pointer to top left element of matrix A
+    /// @param uplo specifies whether the matrix A is an upper or lower triangular
+    /// @param diag specifies whether or not A is unit triangular
+    /// @param C pointer to top left element of matrix C
+    ///
+    template <typename ST, typename MPB, typename MPA, typename MPC>
+    requires MatrixPointer<MPB, ST> && MatrixPointer<MPA, ST> && MatrixPointer<MPC, ST>
+        && (StorageOrder_v<MPB> == columnMajor) && (StorageOrder_v<MPC> == columnMajor)
+    inline void trmm(size_t M, size_t N, ST alpha, MPB B, MPA A, UpLo uplo, bool diag, MPC C)
     {
+        using ET = ST;
         size_t constexpr TILE_SIZE = TileSize_v<ET>;
 
-        size_t const M = rows(B);
-        size_t const N = columns(B);
+        if (diag)
+            BLAST_THROW_EXCEPTION(std::logic_error {"Unit-triangular matrices support not implemented in trmm()"});
 
-        if (rows(A) != N || columns(A) != N)
-            BLAZE_THROW_INVALID_ARGUMENT("Matrix sizes do not match");
+        if (uplo == UpLo::Lower)
+        {
+            size_t j = 0;
 
-        if (rows(C) != M || columns(C) != N)
-            BLAZE_THROW_INVALID_ARGUMENT("Matrix sizes do not match");
+            // Main part
+            for (; j + TILE_SIZE <= N; j += TILE_SIZE)
+            {
+                // size_t const K = N - j - TILE_SIZE;
+                size_t i = 0;
 
-        size_t j = 0;
+                // i + 4 * TILE_SIZE != M is to improve performance in case when the remaining number of rows is 4 * TILE_SIZE:
+                // it is more efficient to apply 2 * TILE_SIZE kernel 2 times than 3 * TILE_SIZE + 1 * TILE_SIZE kernel.
+                for (; i + 3 * TILE_SIZE <= M && i + 4 * TILE_SIZE != M; i += 3 * TILE_SIZE)
+                {
+                    RegisterMatrix<ET, 3 * TILE_SIZE, TILE_SIZE, columnMajor> ker;
+                    gemm(ker, N - j, alpha, B(i, j), A(j, j));
+                    /*
+                    ker.trmmRightLower(alpha, ptr<aligned>(B, i, j), ptr<aligned>(A, j, j));
+                    ker.gemm(K, alpha, ptr<aligned>(B, i, j + TILE_SIZE), ptr<aligned>(A, j + TILE_SIZE, j));
+                    */
```

Comment:
Is this a 'TODO'? Or what is the purpose of the commented-out code?

Reply:
Similar below

Reply:
I believe the two versions differ in performance. The commented one performs in-register `trmm()` for diagonal blocks, the uncommented does not. We should check which one is faster.

Reply:
Please add this as a TODO then

Reply:
:+1: